File: ts-cancel-service/src/main/java/cancel/service/FeatureFlagService.java
package cancel.service;

import dev.openfeature.sdk.Client;
import dev.openfeature.sdk.FlagEvaluationDetails;
import dev.openfeature.sdk.OpenFeatureAPI;
import org.springframework.stereotype.Service;

import javax.annotation.PostConstruct;

@Service
public class FeatureFlagService {

    private Client client;

    @PostConstruct
    public void initialize() {
        try {
            // Get a named client for cancel-service (like Python version uses "voucher-service")
            this.client = OpenFeatureAPI.getInstance().getClient("cancel-service");
        } catch (Exception e) {
            // silently ignore
        }
    }

    public boolean isEnabled(String flagName) {
        try {
            if (client == null) {
                return false;
            }

            // Use getBooleanDetails to get detailed information (like Python version)
            FlagEvaluationDetails<Boolean> details = client.getBooleanDetails(flagName, false);

            if ("ERROR".equals(details.getReason())) {
                return false;
            }

            return Boolean.TRUE.equals(details.getValue());

        } catch (Exception e) {
            return false;
        }
    }
}

File: ts-voucher-service/feature_flag_service.py
from openfeature import api
from openfeature.contrib.provider.flagd import FlagdProvider
from openfeature.contrib.provider.flagd.config import ResolverType, CacheType


class FeatureFlagService:
    def __init__(self) -> None:
        provider = FlagdProvider(
            resolver_type=ResolverType.RPC,
            host="flagd",
            port=8013,
            cache=CacheType.DISABLED,
            max_cache_size=0,
        )
        api.set_provider(provider)
        self.client = api.get_client("voucher-service")

    def is_enabled(self, flag_name: str) -> bool:
        try:
            details = self.client.get_boolean_details(flag_name, False)

            if getattr(details, "reason", None) == "ERROR":
                return False
            return bool(details.value)
        except Exception:
            return False
